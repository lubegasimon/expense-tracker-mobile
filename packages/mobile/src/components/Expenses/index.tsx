import { Text, View, FlatList, StyleSheet, SafeAreaView } from "react-native";
import AddExpense from "../Expense/AddExpenseButton";
import Error from "../Error/Error";

interface ExpensesProps {
  expenses: expense[];
  error: string;
}

export type expense = {
  id: string;
  name: string;
  amount: number;
  details: string;
  category: string;
  categoryId: string;
  createdAt: string;
};

function EmptyList() {
  return (
    <View>
      <View
        style={{
          alignItems: "center",
          padding: 50,
        }}
      >
        <Text style={{ fontSize: 30 }}>No expenses yet today</Text>
      </View>
    </View>
  );
}

function Expenses({ expenses, error }: ExpensesProps) {
  return (
    <SafeAreaView>
      <View style={styles.container}>
        <View style={styles.row}>
          <Text style={styles.header}>Name</Text>
          <Text style={styles.header}>Amount</Text>
        </View>
        {error ? (
          <Error error={error} />
        ) : (
          <FlatList
            data={expenses}
            keyExtractor={(item) => item.id}
            renderItem={({ item }) => {
              const { name, amount } = item;
              return (
                <View style={styles.row}>
                  <Text style={styles.cell}>{name}</Text>
                  <Text style={styles.cell}>{amount}</Text>
                </View>
              );
            }}
            ListEmptyComponent={EmptyList}
          />
        )}
      </View>
      <View style={{ paddingLeft: 10 }}>
        <AddExpense />
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: { paddingLeft: 10, height: "90%" },
  row: {
    flexDirection: "row",
    justifyContent: "space-between",
    paddingVertical: 8,
    borderBottomWidth: 1,
    borderBottomColor: "#ccc",
  },
  header: { fontWeight: "bold", fontSize: 16, flex: 1 },
  cell: { fontSize: 14, flex: 1 },
});

export default Expenses;
