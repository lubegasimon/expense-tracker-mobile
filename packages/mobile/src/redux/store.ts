import { configureStore } from "@reduxjs/toolkit";
import userReducer from "./userSlice";
import expenseCountReducer from "./expenseCountSlice";

export const store = configureStore({
  reducer: {
    user: userReducer,
    expenseCount: expenseCountReducer,
  },
});

type AppStore = typeof store;
export type AppDispatch = AppStore["dispatch"];
export type RootState = ReturnType<AppStore["getState"]>;
