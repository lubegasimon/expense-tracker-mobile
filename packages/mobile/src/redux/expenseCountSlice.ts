import { createSlice } from "@reduxjs/toolkit";
import { RootState } from "./store";

interface ExpenseCountState {
  expenseCount: number;
}

const initialState: ExpenseCountState = {
  expenseCount: 0,
};

const expenseCountSlice = createSlice({
  name: "expenseCount",
  initialState,
  reducers: {
    incrementExpenseCount: (state) => {
      state.expenseCount++;
    },
    decrementExpenseCount: (state) => {
      state.expenseCount--;
    },
  },
});

export const { incrementExpenseCount, decrementExpenseCount } =
  expenseCountSlice.actions;
export const selectExpenseCount = (state: RootState) =>
  state.expenseCount.expenseCount;
export default expenseCountSlice.reducer;
