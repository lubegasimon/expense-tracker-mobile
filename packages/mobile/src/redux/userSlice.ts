import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "./store";

interface UserState {
  email: string;
}

const initialState: UserState = {
  email: "",
};

const userSlice = createSlice({
  name: "user",
  initialState,
  reducers: {
    storedEmail: (state, action: PayloadAction<string>) => {
      state.email = action.payload;
    },
    clearEmail: (state) => {
      state.email = "";
    },
  },
});

export const { storedEmail, clearEmail } = userSlice.actions;
export default userSlice.reducer;
export const selectEmail = (state: RootState) => state.user.email;
