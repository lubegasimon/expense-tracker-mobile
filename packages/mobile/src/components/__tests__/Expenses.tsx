import { render, screen } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
import { Provider } from "react-redux";
import { store } from "@/src/redux/store";
import Expenses from "../Expenses";

/* Hack: Supress React Native animation helpers that scedule timers that conflict with jest */
jest.useFakeTimers();

const mockExpenses = [
  {
    id: "01",
    name: "Water",
    amount: 20,
    details: "Water bills for Feb",
    category: "Utilities",
    categoryId: "",
    createdAt: "12/28/2025",
  },
  {
    id: "02",
    name: "Groceries",
    amount: 50,
    details: "Weekly groceries",
    category: "Food",
    categoryId: "",
    createdAt: "12/28/2025",
  },
];

describe("<Expenses Component/>", () => {
  const mockProps = {
    expenses: mockExpenses,
    error: "",
  };

  it("renders the header correctly", () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <Expenses {...mockProps} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getByText("Name")).toBeTruthy();
    expect(screen.getByText("Amount")).toBeTruthy();
  });

  it("should render a list of expenses", () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <Expenses {...mockProps} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getByText("Water")).toBeTruthy();
    expect(screen.getByText("20")).toBeTruthy();
    expect(screen.getByText("Groceries")).toBeTruthy();
    expect(screen.getByText("50")).toBeTruthy();
  });

  it("should render the empty state when there are no expenses", () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <Expenses expenses={[]} error="" />
        </NavigationContainer>
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getByText("No expenses yet today")).toBeTruthy();
  });

  it("should render an error message when there is an error", () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <Expenses expenses={mockExpenses} error="Something went wrong" />
        </NavigationContainer>
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getByText("Something went wrong")).toBeTruthy();
  });
});
