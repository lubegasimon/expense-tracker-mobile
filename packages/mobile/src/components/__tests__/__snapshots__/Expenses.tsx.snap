// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Expenses Component/> renders the header correctly 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "height": "90%",
        "padding": 10,
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#ccc",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingVertical": 8,
        }
      }
    >
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Name
      </Text>
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amount
      </Text>
    </View>
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={
        [
          {
            "amount": 20,
            "category": "Utilities",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Water bills for Feb",
            "id": "01",
            "name": "Water",
          },
          {
            "amount": 50,
            "category": "Food",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Weekly groceries",
            "id": "02",
            "name": "Groceries",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Water
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              20
            </Text>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Groceries
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              50
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</RCTSafeAreaView>
`;

exports[`<Expenses Component/> should render a list of expenses 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "height": "90%",
        "padding": 10,
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#ccc",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingVertical": 8,
        }
      }
    >
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Name
      </Text>
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amount
      </Text>
    </View>
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={
        [
          {
            "amount": 20,
            "category": "Utilities",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Water bills for Feb",
            "id": "01",
            "name": "Water",
          },
          {
            "amount": 50,
            "category": "Food",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Weekly groceries",
            "id": "02",
            "name": "Groceries",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Water
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              20
            </Text>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Groceries
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              50
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</RCTSafeAreaView>
`;

exports[`<Expenses Component/> should render an error message when there is an error 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "height": "90%",
        "padding": 10,
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#ccc",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingVertical": 8,
        }
      }
    >
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Name
      </Text>
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amount
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "red",
            "margin": 5,
            "marginTop": -5,
            "textAlign": "left",
            "width": 300,
          },
          undefined,
        ]
      }
    >
      Something went wrong
    </Text>
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={
        [
          {
            "amount": 20,
            "category": "Utilities",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Water bills for Feb",
            "id": "01",
            "name": "Water",
          },
          {
            "amount": 50,
            "category": "Food",
            "categoryId": "",
            "createdAt": "12/28/2025",
            "details": "Weekly groceries",
            "id": "02",
            "name": "Groceries",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Water
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              20
            </Text>
          </View>
        </View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "borderBottomColor": "#ccc",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingVertical": 8,
              }
            }
          >
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              Groceries
            </Text>
            <Text
              style={
                {
                  "flex": 1,
                  "fontSize": 14,
                }
              }
            >
              50
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</RCTSafeAreaView>
`;

exports[`<Expenses Component/> should render the empty state when there are no expenses 1`] = `
<RCTSafeAreaView>
  <View
    style={
      {
        "height": "90%",
        "padding": 10,
      }
    }
  >
    <View
      style={
        {
          "borderBottomColor": "#ccc",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingVertical": 8,
        }
      }
    >
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Name
      </Text>
      <Text
        style={
          {
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Amount
      </Text>
    </View>
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={[]}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View>
          <View
            style={
              {
                "alignItems": "center",
                "padding": 50,
              }
            }
          >
            <Text
              style={
                {
                  "fontSize": 30,
                }
              }
            >
              No expenses yet
            </Text>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</RCTSafeAreaView>
`;
